lib/logstash/inputs/rabbitmq.rb

```diff
@@ -131,17 +139,87 @@ def consume!
         end
       end
 
+      public
       def stop
         super
         shutdown_consumer
         close_connection
       end
 
+      public
       def shutdown_consumer
         return unless @consumer
         @consumer.gracefully_shut_down
       end
 
+      private
+      def get_header_value(value)
+        # Two kinds of values require exceptional treatment:
+        #
+        # String values are instances of
+        # com.rabbitmq.client.impl.LongStringHelper.ByteArrayLongString
+        # and we don't want to propagate those.
+        #
+        # List values are java.util.ArrayList objects and we need to
+        # recurse into them to convert any nested strings values.
+        if value.respond_to? "toArray"
+          value.map{|item| get_header_value(item) }
+        elsif value.respond_to? "toString"
+          value.toString
+        else
+          value
+        end
+      end
+
+      private
+      def get_headers(metadata)
+        if not metadata.headers.nil?
+          Hash[metadata.headers.map {|k, v| [k, get_header_value(v)]}]
```

Comment:
Sorry, my limited Ruby-fu isn't following along. Isn't this a plain old Hash constructor? I thought lazy evaluation only took place with a block constructor.

```
$ irb
jruby-1.7.23 :002 > Hash[{"a" => "b", "c" => "d"}.map {|k, v| [k, v.upcase]}].keys
 => ["a", "c"]
```